keps/sig-storage/1790-recover-resize-failure/README.md

```diff
@@ -288,73 +287,70 @@ after expansion is complete even with older kubelets. No recovery from expansion
 
 ### Rollout, Upgrade and Rollback Planning
 
-_This section must be completed when targeting beta graduation to a release._
-
 * **How can a rollout fail? Can it impact already running workloads?**
   This change should not impact existing workloads and requires user interaction via reducing pvc capacity.
 
 * **What specific metrics should inform a rollback?**
+  No specific metric but if expansion of PVCs are being stuck (can be verified from `pvc.Status.Conditions`)
+  then user should plan a rollback.
 
 * **Were upgrade and rollback tested? Was upgrade->downgrade->upgrade path tested?**
-  Describe manual testing that was done and the outcomes.
-  Longer term, we may want to require automated upgrade/rollback tests, but we
-  are missing a bunch of machinery and tooling and do that now.
+  We have not fully tested upgrade and rollback but as part of beta process we will have it tested.
 
 * **Is the rollout accompanied by any deprecations and/or removals of features,
   APIs, fields of API types, flags, etc.?**
-  Even if applying deprecation policies, they may still surprise some users.
+  This feature deprecates no existing functionality.
 
 ### Monitoring requirements
 
 _This section must be completed when targeting beta graduation to a release._
 
 * **How can an operator determine if the feature is in use by workloads?**
-  Ideally, this should be a metrics. Operations against Kubernetes API (e.g.
-  checking if there are objects with field X set) may be last resort. Avoid
-  logs or events for this purpose.
-
+  Any volume that has been recovered will emit a metric: `operation_operation_volume_recovery_total{state='success', volume_name='pvc-abce'}`.
+  
 * **What are the SLIs (Service Level Indicators) an operator can use to
   determine the health of the service?**
   - [ ] Metrics
-    - Metric name:
-    - [Optional] Aggregation method:
-    - Components exposing the metric:
+    - controller expansion operation duration:
```

Comment:
I have a feeling that having the name of the operation as a label will make the metrics harder to use compared to if we had metrics dedicated for each operation such as `storage_operation_expand_volume_duration_seconds`.
From a user perspective, I think that it will be harder to know what are the different types of operations on which SLIs can be computed.

Do you perhaps have a list of all the possible operations so that I can put it into perspective?

Reply:
These metrics are already there and have been there for a long time (like may be 1.11 release and earlier).

Reply:
Ok

Reply:
Looking for OperationCompleteHook in pkg/volume it looks like we have about 14. I think it's been a lot easier to manage with a single metric with a status field---our slo processing within google, for instance, would be a lot more toil-y if we had to pull a metric per operation.

"verify_volumes_are_attached_per_node"
"verify_volumes_are_attached"
"volume_attach"
DetachOperationName
"volume_mount"
"volume_unmount"
"unmount_device"
"map_volume"
"unmap_volume"
"unmap_device"
"verify_controller_attached_volume"
"expand_volume"
"expand_volume"
"volume_fs_resize"

Reply:
Yeah since there are a lot of possible operations, it should be easier to have all of them under the same metric. Just wanted to check if that was the case here or not.

>  I think it's been a lot easier to manage with a single metric with a status field

The expression should be very similar whether we have one or multiple metrics. The difference is that fitting a lot of information in a histogram is very expensive and tends to be counterintuitive to users who would expect counter metrics to also be present and to have more granularity if necessary.

Also when looking into the duration of your operations, the status information is superflux so we will be exposing it in all the buckets of the histograms even though it will not be useful.

Reply:
> The expression should be very similar whether we have one or multiple metrics. The difference is that fitting a lot of information in a histogram is very expensive and tends to be counterintuitive to users who would expect counter metrics to also be present and to have more granularity if necessary.

oh, that's an interesting point.

> 
> Also when looking into the duration of your operations, the status information is superflux so we will be exposing it in all the buckets of the histograms even though it will not be useful.

I'm not sure about this point---if, eg, certain errors have higher latency that might be an interesting thing to know?

Reply:
> I'm not sure about this point---if, eg, certain errors have higher latency that might be an interesting thing to know?

I would agree that it may be useful under certain circumstances, but in the majority of cases, you only care about the average and the 99th percentile when using the histogram and these operations tend to not be impacted by errors since they are uncommon and a majority of them is returning earlier than normal code paths. As such, considering the cost of this information compared to its value, I think logs are better suited since they would make this information cheaper.

Also for SLOs, errors are already counted as part of the unavailability so you don't need to define duration thresholds for them.

I think the choice of what to put into histograms really depends on the scenario, but for Kubernetes in particular where we try to keep the number of metrics exposed under control, I think the cost vs value of adding the status label isn't necessarily worth it to have in the histogram metric.